Generate IR definitions by annotation processor - 1st step

Akirathan opened the pull request:
This PR is the initial step for migrating all the IR elements implemented as Scala case classes to Java classes generated by an annotation processor.

This PR was created from the abandoned #11267 that tried to generated subclasses.

### Pull Request Description

The overall description is the same as in #11267, but this approach tries to generate super classes as suggested in https://github.com/enso-org/enso/pull/11267/files#r1869342527

### References
- New project `runtime-parser-dsl` contains new annotations for IR elements. The docs for these annotations are the most important. See [org.enso.runtime.parser.dsl package](https://github.com/enso-org/enso/blob/ac0d78bc84af28bab072cba5fb58d8e61caebe40/engine/runtime-parser-dsl/src/main/java/org/enso/runtime/parser/dsl)
- New project `runtime-parser-processor` contains the implementation of the annotation processor. The entry point to annotation processing is in [org.enso.runtime.parser.processor.IRProcessor](https://github.com/enso-org/enso/blob/bd4b0b2f99d2490508e7f2fae9e2e07452b0ea77/engine/runtime-parser-processor/src/main/java/org/enso/runtime/parser/processor/IRProcessor.java).
- New project `runtime-parser-processor-tests` depends on aforementioned projects as well as on `runtime-parser`. 
  - The main test is [TestIRProcessorInline.java](https://github.com/enso-org/enso/blob/3086f52b106558da0af7bd47f2ee3eeb77c99737/engine/runtime-parser-processor-tests/src/test/java/org/enso/runtime/parser/processor/test/TestIRProcessorInline.java) that tests the annotation processor by compiling snippets of code with the help of [com.google.testing.compile:compile-testing library](https://github.com/enso-org/enso/blob/0edd8324c51306cb20b12cae5f7223a97e678ce4/build.sbt#L3284).
  - Another test is [GeneratedIRTest.scala](https://github.com/enso-org/enso/blob/e2c3c3bce8070be99ce5fe4f85dd6b4c068fdc00/engine/runtime-parser-processor-tests/src/test/scala/org/enso/runtime/parser/processor/test/GeneratedIRTest.scala) that test the usage of generated code.

### Important Notes

Apart from the annotation processor implementation and tests, [CallArgument.Specified](https://github.com/enso-org/enso/blob/80509a1f2420e97b3879fc8e925a9a863373826d/engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/CallArgument.java) was migrated to the new approach from Scala case class. I uploaded its generated class code in [this gist](https://gist.github.com/Akirathan/0e43de77fe91da399406c1ee7ac2cecd) so you can see it without compilation.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] Benchmarks are OK
- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

Commit:
Add parser-dsl and parser-processor projects

Commit:
Add JModule

Commit:
Fix module declaration - no compiler module dependency

Commit:
Add annotation processor testing.

Commit:
Rewrite some IR elements to IRNode annotated interfaces

Commit:
IRProcessor does some sanity checks

Commit:
IRNode does not have name parameter

Commit:
IRProcessor generates source file for record

Commit:
Add imports and overriden IR methods to generated record

Commit:
Add field processing

Commit:
IRProcessor generates classes, not records

Commit:
IRProcessor generates overrides for user-defined parameterless methods

Commit:
Add Field.isExpression method

This will be important for mapExpressions method implementation

Commit:
sbt picks up Scala case classes generated by Java annotation processor

Commit:
IRProcessor generates Scala classes.

This is the only way to retain 100% backward compatibility.

Commit:
Revert "sbt picks up Scala case classes generated by Java annotation processor"

This reverts commit d2bb8c354a8d9b409c9e6ceb7fb1f788740721aa.

Commit:
Revert "IRProcessor generates Scala classes."

This reverts commit 68a070b98b03528cb42f5788bbb2cb6915ece2ad.

Commit:
IRProcessor handles primitive fields

Commit:
Remove unused methods

Commit:
IRProcessor fails fast

Commit:
IRProcessor can collect fields from super interfaces

Commit:
Fix collecting fields from super interfaces.

Commit:
Prepare IRNodeProcessor for processing multiple nested interfaces

Commit:
IRNodeProcessor handles multiple nested interfaces

Commit:
runtime-parser-processor does not depend on runtime-parser

Commit:
Move parser processor tests to a separate project

Commit:
runtime-parser-processor does not depend on runtime-parser

Commit:
runtime-parser depends on runtime-parser-processor

Commit:
IRProcessor adds package

Commit:
ParserDependenciesTest ensures that IRNodeProcessor is on class path

Commit:
runtime-parser-processor module provides annotation processor service

Commit:
Fix getSimpleTypeName in ReferenceField

Commit:
No nested @IRNode annotation

Commit:
JModule extends IR

Commit:
IRNodeProcessor handles parametrized List child

Commit:
runtime-parser-processor-tests runs with enabled assertions

Commit:
Test IRNode with List of children

Commit:
All IR interfaces use scala list, not java.util.List

Commit:
Do not process already overriden methods

Commit:
Enable IRProcessor in runtime-parser

Commit:
Ensure that only fields from current interface and its super interfaces are collected.

No methods from nested interfaces.

Commit:
Implement passData, location and diagnostics methods

Commit:
Override duplicate method

Commit:
Fix duplicate method generation

Commit:
Move duplicate method generation to a separate class

Commit:
Duplicate method duplicates also non-child fields

Commit:
Simplify tests

Commit:
During runtime-parser-processor-test compilation, IRNodeProcessor runs

Commit:
docs

Commit:
Add some IR interfaces to the testing code

Commit:
test generated list children

Commit:
DuplicateMethodGenerator duplicates also meta fields

Commit:
Builder allows setting meta fields

Commit:
Add tests for duplicated metadata

Commit:
Add runtime-parser-processor-tests to the enso aggregate

Commit:
Add IRCopyMethod annotation

Commit:
Iteration over super interfaces is handled only via SuperInterfaceVisitor

Commit:
Simplify super interface iteration

Commit:
Builder has a copy constructor

Commit:
Implement CopyMethodGenerator

Commit:
Fix all tests

Commit:
Test copy method generation

Commit:
Do not try to override static or default methods

Commit:
docs

Commit:
Allow multiple copy methods

Commit:
CopyMethodGenerator generates copy method with correct order of parameters

Commit:
Test valid parameter names for copy method

Commit:
More tests

Commit:
Generate mapExpressions method

Commit:
Add mock definition of JCallArgument

Commit:
Use qualified type names in mapExpressions method

Commit:
Implement JBlank.create factory method

Commit:
JExpression overrides duplicate and mapExpressions

Commit:
Merge branch 'develop' into wip/akirathan/java-ir

Commit:
Remove unused import

Commit:
Move all the sample `JXyz` interfaces to tests

Commit:
Generate identifiedLocation method override

Commit:
Fix compilation after the move

Commit:
Move sample `JXyz` interfaces to a dedicated test package

Commit:
Add sample definition of JEmpty

Commit:
Replace Scala Empty with Java Empty annotated with @IRNode

Commit:
Refactor Utils.findMethod

Commit:
Fix test after merge of develop

Commit:
Fix TestIRProcessorInline after `J*` classes were moved to test dir

Commit:
Implement SetLocationMethodGenerator

Commit:
fmt

Commit:
Ensure only a single IR interface is extended

Commit:
Fix compilation after Refactoring Empty to an interface with @IRNode

Commit:
Empty extends Expression

Commit:
Override equals and hashCode

Commit:
Explicitly list all the processor classes in javacOptions

Commit:
Refactor field class to field package

Commit:
Refactor method generation classes to methodgen package

Commit:
Introduce utils package

Commit:
Add GenerateIR annotation

Commit:
Remove tests using the IRNode annotation

Commit:
Revert "Empty extends Expression"

This reverts commit 4a0ea14a3752dcc149c704f856059a2368cf7ad2.

Commit:
Revert "Fix compilation after Refactoring Empty to an interface with @IRNode"

This reverts commit 5ac716a01517cb12dd919a96271d4932ade714f9.

Commit:
Revert "Replace Scala Empty with Java Empty annotated with @IRNode"

This reverts commit aa345ed456cbc05db2baed83bc4b034e973ceeb6.

Commit:
Remove all test interfaces using IRNode annotation

Commit:
Remove TestGeneratedIR

Commit:
Update TestIRPRocessorInline.

Remove IRNode annotation

Commit:
GenerateIR uses IR as the default interface

Commit:
Skeleton of implementation

Commit:
Remove IRNode

Commit:
Better error msg in FieldCollector

Commit:
Add import in test

Commit:
Ensure GenerateIR.interfaces is subtype of IR

Commit:
Add import in test

Commit:
Fix generic type introspection from VariableElement

Commit:
Fix imports in rest of the tests.

TestIRProcessorInline are now green.

Commit:
Generated constructor is protected

Commit:
Generate second constructor - with user fields

Commit:
style - add new line

Commit:
Add docs to the generated ctors.

Commit:
Add sample JExpression to tests

Commit:
fmt

Akirathan commented:
As of https://github.com/enso-org/enso/pull/11770/commits/b8f6b8f21735f3c1302118e94a005aede07e7556 all tests in `runtime-parser-processor-tests` passes.

An example of generated code for [JExpression](https://github.com/enso-org/enso/blob/0ea630d66ada5d80c5a8fdc942efef3a51706f8f/engine/runtime-parser-processor-tests/src/main/java/org/enso/runtime/parser/processor/test/gen/ir/core/JExpression.java) is:
- [JBlockGen](https://gist.github.com/Akirathan/21ffa5e8c758a7857fd63fff1cc69ee7#file-jblockgen-java)
- [JBindingGen](https://gist.github.com/Akirathan/21ffa5e8c758a7857fd63fff1cc69ee7#file-jbindinggen-java)

Commit:
Add java docs to the generated class

JaroslavTulach commented:
> * [JBlockGen](https://gist.github.com/Akirathan/21ffa5e8c758a7857fd63fff1cc69ee7#file-jblockgen-java)

I think the generated code looks reasonably nice, but:
- when serializing, we do not want to intialize `passData()` neither `diagnostics()` if they were not initialized before
- probably the `@Persistance` annotation shall be in the same package and have _package private_ access to these special fields
- `JBindingGen` should be `abstract`
- builder should return `JBinding` and not the generated class
- I'd like to remind you of the [WhiningBuilder pattern](http://wiki.apidesign.org/wiki/WhiningBuilder) - better than `validate()` as it does compile type checks
- the `build()` method shall `throw NameExc, ExpressionExc` and the builder type should be `Builder<NameExc extends Exception, ExpressionExc extends Exception>`
- empty builder is initialized with as `Builder<MissingNameException, MissingExpressionException>`
- Signature of name method is: `Builder<RuntimeException, ExpressionExc> name(String name)` - to signal that _name has already been provided_
- `name.duplicate(...)` should be overridden in `Name` to return `Name`
- `IR setLocation(Option<IdentifiedLocation> location)` should return `JBinding`
- `JExpression duplicate(` should return `JBinding`
- `showCode` may be implemented manually in `JBinding` - another reason for making `JBindingGen` `abstract`
- missing new line in `list.add(returnValue);return scala.jdk.javaapi.CollectionConverters.asScala(list).toList();`

As a concept it is nice, but the real challenge is going to be the implementation of real `IR` element(s).

Commit:
Remove unused IRCopyMethod

Commit:
Replace scala Empty with generated IR

Commit:
Support scala.Option field

Commit:
Add nl

JaroslavTulach commented:
I see [one failure](https://github.com/enso-org/enso/actions/runs/12263203676/job/34214229485?pr=11770#step:7:4841) related to `Empty` and `EmptyGen`. Are there any other important failures?

Commit:
Add JCallArgument to tests

Commit:
Merge branch 'develop' into wip/akirathan/java-ir-superclass

Commit:
Test that annotated constructor can have meta parameter

Commit:
Constructor parameters must be annotated

Commit:
Constructor parameters can be meta type

Commit:
Generated super class is abstract

Commit:
Annotated class must extend the generated super class

Commit:
Fix tests - annotated class must extend the generated super class

Commit:
Use specific IRProcessingException.

- Do not use asserts or any other exceptions.
- Error is printed ony in the main processing loop.

Commit:
setLocation returns subclass type

Commit:
Reorder generated fields in JCallArgument

Commit:
mapExpressions returns subclass type

Commit:
Add scala tests to runtime-parser-processor-tests

Commit:
[WIP] Demonstrate implementation of unapply

Akirathan commented:
## Unapply & apply
In order to make necessary modifications in `runtime-parser` as minimal as possible, we *could* implement `unapply` methods on Java classes to make it work with Scala deconstruct pattern matching.

0a1f5a88695c8f1c359e45eb72bdce94215e23aa demonstrates how a Java class `JSpecified` can implement public static `unapply` method so it works in Scala deconstruct match. The Scala compiler is happy and the tests succeeds. But the IDE is not happy:
![image](https://github.com/user-attachments/assets/4fe360e9-684c-451b-b000-55fdce6d7a64)

I assume VSCode with Metals extension will also not be able to deduce that `JSpecified` java class _can_ be treated as a case class in this particular case.

The question is: is this really worth it? Wouldn't it be better to just incrementally rewrite pattern matching in Scala that uses deconstruct to simpler variants that just matches on types? For example:
```scala
    val callArg = new JSpecified(true, None, lit)
    callArg match {
      case JSpecified(isSynthetic, _, _) =>
        isSynthetic shouldEqual true
    }
```
Could be rewritten to
```scala
    callArg match {
      case specified: JSpecified =>
        specified.isSynthetic shouldEqual true
    }
```
without any behavioral changes. Moreover, the latter is OK in IDE.

I would like to first create a PR that simplifies pattern matching, for `CallArgument`, and then continue with the migration. Any strong opinions against @JaroslavTulach @hubertp ?

### TL;DR;
Don't try to bend Java classes to look like Scala case classes. Just refactor deconstructing pattern matches in Scala to simpler variants.

JaroslavTulach commented:
I don't mind how this particular IDE behaves. All that matters is whether the code compiles and behaves as it should.

Akirathan commented:
> I don't mind how this particular IDE behaves. All that matters is whether the code compiles and behaves as it should.

One of our long-term visions is to replace all Scala code in `runtime-parser` with Java. Refactoring IR elements such that Scala compiler does not use its magical `apply` and `unapply` methods (as demonstrated in https://github.com/enso-org/enso/pull/11943) only helps this goal. Such refactoring is trivial. Not only it moves us closer to Java, but it also makes the IJ IDE (and presumably also VSCode with Metals) experience much better. Imagine a situation where half of our IR elements are migrated to generated Java classes - that would make almost half of the code base in `runtime-parser` unreadable in the IDE and the experience would be almost the same as using plain text editor. If the "fix" is so straightforward, as #11943, I am strongly pushing for it. I would not like to decrease the DevX anymore, as it already is quite low.

JaroslavTulach commented:
> presumably also VSCode with Metals

I doubt. Metals is using completely different technology than your IDE.

Commit:
Revert "[WIP] Demonstrate implementation of unapply"

This reverts commit 0a1f5a88695c8f1c359e45eb72bdce94215e23aa.

Commit:
Merge branch 'develop' into wip/akirathan/java-ir-superclass

# Conflicts:
#	engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/pass/desugar/OperatorToFunctionTest.scala

JaroslavTulach commented on engine/runtime-parser-dsl/src/main/java/org/enso/runtime/parser/dsl/GenerateIR.java:
```diff
@@ -0,0 +1,31 @@
+/**
+ * A class annotated with this annotation will be processed by the IR processor. The processor will
+ * generate a super class that is meant to be extended by this class. The generated class will have
+ * the same package as this class, and its name will have the "Gen" suffix. Majority of the methods
```
It is possible to extract the name of the super class from `class X extends MyParent` definition. E.g. then the name of the super class could be arbitrary. It would have to be in the same package, but there would be no restriction on the `Gen` suffix.

Akirathan replied:
Implemented and tested in bdd73c351ab1b778a92ce279e7be0596f8bda6f3

JaroslavTulach commented on engine/runtime-parser-dsl/src/main/java/org/enso/runtime/parser/dsl/GenerateIR.java:
```diff
@@ -0,0 +1,31 @@
+   *
+   * @return
+   */
+  String interfaces() default "org.enso.compiler.core.IR";
```
I guess this is not used yet. I see plural being used here. In such case the value should be `[]` array. I would also suggest to use `Class[]` that way one prevents typos as the referenced classes must exist. Btw. looking at `Block` I see it implements `IRKind.Primitive` - that interface isn't subtype of `IR`...

Akirathan replied:
`GenerateIR.interfaces` refactored to `Class[]` in e54fdb933026929d618c3855689657ef57059d98.

Akirathan replied:
In 9da9394508cb519b6035d559d4230e773ef9bab1, one can specify multiple `interfaces`. Moreover, I have updated the docs and tests. There is no need for the `interfaces` to be subtypes of `org.enso.compiler.core.IR` anymore.

JaroslavTulach commented on engine/runtime-parser-dsl/src/main/java/org/enso/runtime/parser/dsl/IRChild.java:
```diff
@@ -0,0 +1,19 @@
+
+/**
+ * Constructor parameter annotated with this annotation will be represented as a child field in the
+ * generated super class. Children of IR elements for a tree. A child will be part of the methods
```
```suggestion
 * generated super class. Children of IR elements form a tree. A child will be part of the methods
```

Akirathan replied:
Typo fixed in ac0d78bc84af28bab072cba5fb58d8e61caebe40

JaroslavTulach commented on engine/runtime-parser-dsl/src/main/java/org/enso/runtime/parser/dsl/IRChild.java:
```diff
@@ -0,0 +1,19 @@
+ * Constructor parameter annotated with this annotation will be represented as a child field in the
+ * generated super class. Children of IR elements for a tree. A child will be part of the methods
+ * traversing the tree, like {@code mapExpression} and {@code children}. The parameter type must be
+ * a subtype of {@code org.enso.compiler.ir.IR}.
```
```suggestion
 * a subtype of {@link org.enso.compiler.ir.IR}.
```

Akirathan replied:
Cannot use `link` here, because `runtime-parser-dsl` project does not depend on `runtime-parser`.

JaroslavTulach commented on engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/Empty.java:
```diff
@@ -0,0 +1,16 @@
+import org.enso.runtime.parser.dsl.GenerateFields;
+import org.enso.runtime.parser.dsl.GenerateIR;
+
+@GenerateIR(interfaces = "org.enso.compiler.core.ir.Expression")
```
Using `Class[]` might be better. Then this line would become:
```suggestion
@GenerateIR(interfaces =Expression.class)
```

Akirathan replied:
Already implemented. See https://github.com/enso-org/enso/pull/11770#discussion_r1899542105

JaroslavTulach commented on engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/DiagnosticStorage.scala:
```diff
@@ -61,3 +61,7 @@ final class DiagnosticStorage(initDiagnostics: Seq[Diagnostic] = Seq())
 }
+
+object DiagnosticStorage {
+  def empty(): DiagnosticStorage = new DiagnosticStorage()
```
`create()` or `createEmpty()` might be better to indicate that a new instance of this mutable class is always created.

Akirathan replied:
Renamed in b3ce70420d775f494a45f93d8588df784da5aaa1

JaroslavTulach reviewed (commented):
Few notes.

Commit:
Improve error message in IRProcessor

Commit:
Fix check of super class

Commit:
Generated classes are compiled with -parameters

Commit:
GenerateIR.interfaces is of type Class[]

Commit:
Can specify multiple GenerateIR.interfaces to implement

Commit:
Generated super class can have arbitrary name

Commit:
Fix typo in docs

Commit:
Rename DiagnosticStorage.empty to createEmpty

Commit:
Fix build

Commit:
Duplicate method has specific return type

Commit:
Generate also parameterless duplicate method

Commit:
Generated class can already exist - will be rewritten

Commit:
Generate diagnosticsCopy.

From LazyDiagnostics trait.

Commit:
Replace CallArgument Scala code with the generated class.

Commit:
Fix duplicate method - passData and diagnostics are copied.

Commit:
Generate toString method

Commit:
Fix expected error message

Commit:
FIx ensureExtendsGeneratedSuperclass

Commit:
Fix NPE in duplicate

Commit:
mapExpressions duplicates this iff one of the mapped children changed

Commit:
mapExpressions do not recurse on Expression children

Commit:
Specified.copy only copies if parameter is different

Commit:
Pass null to meta field in super ctor

Commit:
Fix whitespaces in formatting of generated classes

Commit:
No "copy constructor" for Builder

Commit:
imports are sorted

JaroslavTulach commented on engine/runtime-parser/src/main/java/module-info.java:
```diff
@@ -2,6 +2,8 @@
   requires scala.library;
   requires org.enso.persistance;
+  requires static org.enso.runtime.parser.dsl;
+  requires static org.enso.runtime.parser.processor;
```
Requires static is correct.

JaroslavTulach commented on engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/CallArgument.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    public Specified copy(Expression value) {
+      if (value != this.value()) {
```
Looks like boiler plate code. I was hoping this code would be generated.

Akirathan replied:
Copy method generated in 1ceabfeb292f90e87838d3476685834960bc58e3

JaroslavTulach commented on engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/Empty.java:
```diff
@@ -0,0 +1,16 @@
+public final class Empty extends EmptyGen {
+  @GenerateFields
+  public Empty(IdentifiedLocation identifiedLocation, MetadataStorage passData) {
+    super(null, passData, identifiedLocation, null);
```
What does it mean `super(null, ..., null)`? Why two `null` parameters?

Akirathan replied:
[IRNodeClassGenerator](https://github.com/enso-org/enso/blob/80509a1f2420e97b3879fc8e925a9a863373826d/engine/runtime-parser-processor/src/main/java/org/enso/runtime/parser/processor/IRNodeClassGenerator.java#L178-L215) generates two constructors - the "main" constructor which has all user-defined and meta fields as parameters, and another one that has only user-defined fields as parameters. This `super` calls the constructor which has meta fields. There are four meta fields. It is possible, and probably better, to generate constructor in the super class that exactly matches the constructor here though.

Akirathan replied:
Only a single constructor matching subtype's constructor is generated in 0e15befd717070c938a620be2aa99f38414164a4

JaroslavTulach commented on engine/runtime-parser/src/test/java/org/enso/compiler/core/ParserDependenciesTest.java:
```diff
@@ -37,4 +41,14 @@ public void avoidPolyglotDependency() {
+  @Test
+  public void parserProcessorIsAvailable() {
+    try {
+      var clazz = Class.forName("org.enso.runtime.parser.processor.IRProcessor");
```
Strange test. Why are we testing this class can be loaded? Of course, because it on module path of `runtime-parser` project as `provided` and that implies it is visible in the tests of the `runtime-parser` project.

I'd rather have a test in `runtime-integration-tests` that `IRProcessor` class **cannot be loaded**.  Because we don't want this class in the final JARs that we deploy.

JaroslavTulach reviewed (approved):
I am surprised there is hand written implementation of `copy` method in `CallArgument.Specified` - it'd be better if it was generated. The amount of arguments sent to `super` constructor is surprising.

hubertp commented on build.sbt:
```diff
@@ -3218,14 +3229,76 @@ lazy val `runtime-parser` =
         "org.netbeans.api" % "org-openide-util-lookup" % netbeansApiVersion
       ),
+      // Java compiler is not able to correctly find all the annotation processor, because
+      // one of the is on module-path. To overcome this, we explicitly list all of them here.
```
`one of the is on module-path`?

Akirathan replied:
Typo fixed in fab106fbea6075fc515d5d01f40f76041e5b5c01

hubertp commented on engine/runtime-parser-dsl/src/main/java/org/enso/runtime/parser/dsl/GenerateIR.java:
```diff
@@ -0,0 +1,32 @@
+  /**
+   * Interfaces that the generated superclass will implement. The list of the interfaces will simply
+   * be put inside the {@code implements} clause of the generated class. All the generated classes
+   * implements {@code org.enso.compiler.core.IR} by default.
```
```suggestion
   * implement {@code org.enso.compiler.core.IR} by default.
```

hubertp commented on engine/runtime-parser-processor-tests/src/main/java/org/enso/runtime/parser/processor/test/gen/ir/core/JCallArgument.java:
```diff
@@ -0,0 +1,42 @@
+/** Call-site arguments in Enso. */
+public interface JCallArgument extends IR {
+  /** The name of the argument, if present. */
+  Option<Name> name();
```
Why `Scala` option?

Akirathan replied:
To retain backward compatibility. We use `scala.Option` all over the place in IR types.

hubertp commented on engine/runtime-parser-processor-tests/src/test/scala/org/enso/runtime/parser/processor/test/GeneratedIRTest.scala:
```diff
@@ -0,0 +1,37 @@
+  "JSpecifiedGen" should "have generated parameter names with javac compiler" in {
+    val lit     = Literal.Text("foo", null, new MetadataStorage())
+    val callArg = new JSpecified(isSynthetic = true, value = lit, name = None)
+    callArg should not be null
```
that's always guaranteed to be true, no?

Akirathan replied:
It is. This test is here just to ensure that `new JSpecified` can be used with named arguments.

hubertp commented on engine/runtime-parser-processor/src/main/java/org/enso/runtime/parser/processor/GenerateIRAnnotationVisitor.java:
```diff
@@ -0,0 +1,68 @@
+   * Returns a type from {@link GenerateIR#interfaces()} that is a subtype of {@code
+   * org.enso.compiler.core.IR}. There must be only one such subtype specified.
+   *
+   * @return May be null.
```
and if it is `null` then what does it mean?

Akirathan replied:
Docs added in 3d865b9b8e483f4c1f4b387e24cfce8d38a76142

hubertp commented on engine/runtime-parser-processor/src/main/java/org/enso/runtime/parser/processor/GeneratedClassContext.java:
```diff
@@ -0,0 +1,221 @@
+    for (var param : ctor.getParameters()) {
+      var paramSimpleType = simpleTypeName(param);
+      var paramName = param.getSimpleName().toString();
+      Supplier<Void> printErrAndFail =
```
That feels like an overkill. A private method would be enough

Akirathan replied:
Replaced with private method in 412d46626bcead71d5e66ffd749862f57c8889bc

hubertp commented on engine/runtime-parser-processor/src/main/java/org/enso/runtime/parser/processor/GeneratedClassContext.java:
```diff
@@ -0,0 +1,221 @@
+          }
+        };
+    var typeName = paramType.accept(typeVisitor, null);
+    ;
```
```suggestion
```

hubertp commented on engine/runtime-parser-processor/src/main/java/org/enso/runtime/parser/processor/IRNodeClassGenerator.java:
```diff
@@ -0,0 +1,415 @@
+        protected IdentifiedLocation location;
+        protected UUID id;
+        """
+            .replace("$userDefinedFields", userDefinedFields);
```
`StringBuilder`?

Akirathan replied:
I prefer multiline string with `replace` here.

hubertp commented on engine/runtime-parser-processor/src/main/java/org/enso/runtime/parser/processor/methodgen/BuilderMethodGenerator.java:
```diff
@@ -0,0 +1,116 @@
+
+    var code =
+        """
+        public static final class Builder {
```
Could use string templates since we are on latest jdk?

Akirathan replied:
Unfortunately, Frgaal still does not support string templates.

hubertp commented on engine/runtime-parser-processor/src/main/java/org/enso/runtime/parser/processor/methodgen/MapExpressionsMethodGenerator.java:
```diff
@@ -0,0 +1,229 @@
+public final class MapExpressionsMethodGenerator {
+  private final ExecutableElement mapExpressionsMethod;
+  private final GeneratedClassContext ctx;
+  private static final String METHOD_NAME = "mapExpressions";
```
Probably should use came case, as everywhere else

Akirathan replied:
I don't know about "everywhere else". For example, in [Constants.java](https://github.com/enso-org/enso/blob/a862ea7948618dc2812719fe2d5f0a47b8c56d3f/engine/runtime/src/main/java/org/enso/interpreter/Constants.java) we use upper case, as well as in [Builtins.java](https://github.com/enso-org/enso/blob/d687365070381b07eb4214bc665a13912c3bbb96/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java#L74-L76)

hubertp commented on engine/runtime-parser-processor/src/main/java/org/enso/runtime/parser/processor/utils/Utils.java:
```diff
@@ -0,0 +1,215 @@
+              // it.
+              // This is because runtime-parser-processor project does not depend on runtime-parser
+              // and
+              // so the org.enso.compiler.core.IR interface is not available in the classpath.
```
:(

Akirathan replied:
I assume the sad face is there because we have not yet separated the `runtime-parser` project into a separate project that would contain the `IR` interface.

hubertp commented on engine/runtime-parser-processor/src/main/java/org/enso/runtime/parser/processor/utils/Utils.java:
```diff
@@ -0,0 +1,215 @@
+              // This is because runtime-parser-processor project does not depend on runtime-parser
+              // and
+              // so the org.enso.compiler.core.IR interface is not available in the classpath.
+              if (iface.getSimpleName().toString().equals("IR")) {
```
can you make it a constant ?

Akirathan replied:
Constants introduced in 42a988f2aa15c128d6cc591e1fb044c50d7f796c

hubertp commented on engine/runtime-parser-processor/src/main/java/org/enso/runtime/parser/processor/utils/Utils.java:
```diff
@@ -0,0 +1,215 @@
+    return null;
+  }
+
+  public static <T> List<T> minus(List<T> superset, List<T> subset) {
```
I believe in API this is typically called `diff` not `minus`

Akirathan replied:
That seems like better name. Renamed in fe0a378cb2505d61a89e95d3ac54fbbcb37cec9d

hubertp reviewed (commented):
I feel like processor would become much more readable if we used String templates which are now available. A combination of `StringBuilder`s and text replacing can be sometimes hard to read.

hubertp commented:
Is there any impact on benchmarks with all this machinery?

Commit:
Replace Supplier with private method

Commit:
Add docs

Commit:
fix typos

Commit:
Rename minus to diff

Commit:
Introduce constants

Commit:
passData is never null

Commit:
Field uses TypeMirror instead of TypeElement

Commit:
More constant in Utils

Commit:
Move ClassField out of GeneratedClassContext.

Also introduce more fields to ClassField

Commit:
duplicate method compares fqn type names

Commit:
ClassField.canBeNull must be specified in builder

Commit:
Generate validateConstructor method

Commit:
Prefer simple type names in generated source

Commit:
Fix test: passData is never null

Commit:
Fix Utils.simpleTypeName

Commit:
Field is abstract class.

Simplify Field hierarchy

Commit:
Prefer simple type names

Commit:
Generate copy method

Commit:
Generate just a single constructor matching subtype

Commit:
Implement Specified.showCode

Commit:
passData is never null

JaroslavTulach commented on engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/CallArgument.java:
```diff
@@ -0,0 +1,84 @@
+          this.location,
+          this.id,
+          this.name(),
+          value,
```
> Copy method generated in 1ceabfeb292f90e87838d3476685834960bc58e3

Great, this looks like the most elegant code to delegate to.

JaroslavTulach commented:
A problem generating **ScalaDoc** for Java class that extends AP generated class [breaks the CI](https://github.com/enso-org/enso/actions/runs/12602159599/job/35124768999?pr=11770#step:7:5569):
```
[error] /home/runner/work/enso/enso/engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/CallArgument.java:39:33: not found: type SpecifiedGen
[error]   final class Specified extends SpecifiedGen {
[error]                                 ^
[error] /home/runner/work/enso/enso/engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/Empty.java:7:34: not found: type EmptyGen
[error] public final class Empty extends EmptyGen {
[error]                                  ^
[error] two errors found
[info] [FrgaalJavaCompiler] compiling 17 Java sources with module-info.java to /home/runner/work/enso/enso/engine/runtime-parser/target/classes ...
[info] done compiling
[error] Scaladoc generation failed
[error] 	at scala.tools.xsbt.Runner.run(ScaladocBridge.scala:54)
[error] 	at scala.tools.xsbt.ScaladocBridge.run(ScaladocBridge.scala:25)
[error] 	at sbt.internal.inc.AnalyzingCompiler.doc(AnalyzingCompiler.scala:154)
[error] 	at sbt.internal.inc.AnalyzingCompiler.doc(AnalyzingCompiler.scala:133)
[error] 	at sbt.Doc$.$anonfun$scaladoc$1(Doc.scala:53)
[error] 	at sbt.Doc$.$anonfun$scaladoc$1$adapted(Doc.scala:41)
```

Commit:
Generated toString method returns one-line string

Commit:
Update changelog

Commit:
Add generated sources to runtime-parser/Compile/doc/sources.

This fixes publishM2 task

Commit:
Do not link duplicate method in javadoc.

scaladoc fails on that.

Commit:
Run sbt clean before publishing M2

Akirathan commented on .github/workflows/enso4igv.yml:
```diff
@@ -128,6 +128,9 @@ jobs:
           java-version: "21"
           distribution: "graalvm-community"
 
+      - name: Clean sbt
```
Ensure that generated sources are removed before running `sbt publishM2`.

JaroslavTulach replied:
I am afraid `publishM2` will generate them again: https://github.com/enso-org/enso/actions/runs/12634477624/job/35202416216?pr=11770#step:9:5565 and it has to generate all the code otherwise we would be able to use such generated classes in `runtime-parser.jar`, right?

If there is a way to disable ScalaDoc (or replace it with Javadoc) or restrict it somehow, then go that route. We certainly don't need `runtime-parser-javadoc.jar` for anything useful.

Akirathan replied:
I have reverted all the commits changing this workflow file, and "fixed" the `publishM2` task with 5de98f379a3ec90450f367011530e48d328dd2b5 so that src and doc jars are not generated at all.

Akirathan commented:
All the engine tests are green. Scheduling benchmarks:
- engine: https://github.com/enso-org/enso/actions/runs/12632937078
- stdlibs: https://github.com/enso-org/enso/actions/runs/12632946564

<blockquote><img src="https://opengraph.githubassets.com/082eb90989f80c16f6d1d32f4cd3d517a5d39497247f41793e8e046ea04bba0c/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Engine · enso-org/enso@291e593</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Benchmark Engine · enso-org/enso@291e593</div></blockquote>
<blockquote><img src="https://opengraph.githubassets.com/082eb90989f80c16f6d1d32f4cd3d517a5d39497247f41793e8e046ea04bba0c/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Standard Libraries · enso-org/enso@291e593</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Benchmark Standard Libraries · enso-org/enso@291e593</div></blockquote>

Commit:
[WIP] Print sbt config

Commit:
Revert "[WIP] Print sbt config"

This reverts commit dd5efcd2d928673e08116e8ec7c39bc96a0ec660.

Commit:
Revert "Run sbt clean before publishing M2"

This reverts commit 291e593f1802194154de7576ff1b71a199f332be.

Commit:
Revert "Add generated sources to runtime-parser/Compile/doc/sources."

This reverts commit 4226c2e5dd1eec57bfa10dcf3850e58d789a6119.

Commit:
Don't publish doc and src jars to M2

Akirathan commented:
Both engine and stdlib benchmarks are OK. Let's merge this PR ASAP.

Commit:
Add common settings to the projects published to local Maven repo

Commit:
Merge branch 'develop' into wip/akirathan/java-ir-superclass

JaroslavTulach commented on build.sbt:
```diff
@@ -2575,6 +2575,25 @@ lazy val mixedJavaScalaProjectSetting: SettingsDefinition = Seq(
+  */
+lazy val publishLocalSetting: SettingsDefinition = Seq(
+  version := mavenUploadVersion,
+  publish / skip := false,
```
Having shared `publishLocalSetting` is a good idea. Suppressing `packageDoc/publishArtifact` is a neat trick! We wouldn't be allowed to upload the artifact to Maven central repository without sources and javadoc, but for local deployment - it should be good enough solution.

Commit:
toStringMethod converts return value to single line

Commit:
fix: toStringMethod converts return value to single line

Commit:
Merge branch 'develop' into wip/akirathan/java-ir-superclass

# Conflicts:
#	build.sbt

Commit:
Merge branch 'develop' into wip/akirathan/java-ir-superclass
